prow/gcsupload/options.go

```diff
@@ -0,0 +1,122 @@
+*/
+
+package gcsupload
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+
+	"k8s.io/test-infra/prow/kube"
+)
+
+// Options exposes the configuration necessary
+// for defining where in GCS an upload will land.
+type Options struct {
+	// Items are files or directories to upload
+	Items []string `json:"items,omitempty"`
+
+	// SubDir is appended to the GCS path
+	SubDir string `json:"sub_dir,omitempty"`
+
+	kube.GCSConfiguration
+
+	// GcsCredentialsFile is the path to the JSON
+	// credentials for pushing to GCS
+	GcsCredentialsFile string `json:"gcs_credentials_file,omitempty"`
+	DryRun             bool   `json:"dry_run"`
+}
+
+// Complete internalizes command line arguments
+func (o *Options) Complete(args []string) {
+	o.Items = args
+}
+
+// Validate ensures that the set of options are
+// self-consistent and valid
+func (o *Options) Validate() error {
+	if !o.DryRun {
+		if o.Bucket == "" {
+			return errors.New("GCS upload was requested no GCS bucket was provided")
+		}
+
+		if o.GcsCredentialsFile == "" {
+			return errors.New("GCS upload was requested but no GCS credentials file was provided")
+		}
+	}
+
+	if o.PathStrategy != kube.PathStrategyLegacy && o.PathStrategy != kube.PathStrategyExplicit && o.PathStrategy != kube.PathStrategySingle {
+		return fmt.Errorf("GCS path strategy must be one of %q, %q, or %q", kube.PathStrategyLegacy, kube.PathStrategyExplicit, kube.PathStrategySingle)
+	}
+
+	if o.PathStrategy != kube.PathStrategyExplicit && (o.DefaultOrg == "" || o.DefaultRepo == "") {
+		return fmt.Errorf("default org and repo must be provided for GCS strategy %q", o.PathStrategy)
+	}
+
+	return nil
+}
+
+// BindOptions adds flags to the FlagSet that populate
+// the GCS upload options struct given.
+func BindOptions(options *Options, fs *flag.FlagSet) {
+	fs.StringVar(&options.SubDir, "sub-dir", "", "Optional sub-directory of the job's path to which artifacts are uploaded")
+
+	fs.StringVar(&options.PathStrategy, "path-strategy", kube.PathStrategyExplicit, "how to encode org and repo into GCS paths")
+	fs.StringVar(&options.DefaultOrg, "default-org", "", "optional default org for GCS path encoding")
+	fs.StringVar(&options.DefaultRepo, "default-repo", "", "optional default repo for GCS path encoding")
+
+	fs.StringVar(&options.Bucket, "gcs-bucket", "", "GCS bucket to upload into")
+	fs.StringVar(&options.GcsCredentialsFile, "gcs-credentials-file", "", "file where Google Cloud authentication credentials are stored")
+	fs.BoolVar(&options.DryRun, "dry-run", true, "do not interact with GCS")
+}
+
+const (
+	// JSONConfigEnvVar is the environment variable that
+	// utilities expect to find a full JSON configuration
+	// in when run.
+	JSONConfigEnvVar = "GCSUPLOAD_OPTIONS"
+)
+
+// ResolveOptions will resolve the set of options, preferring
+// to use the full JSON configuration variable but falling
+// back to user-provided flags if the variable is not
+// provided.
+func ResolveOptions() (*Options, error) {
```

Comment:
Would it be useful to encapsulate this?

Reply:
In what sense?

Reply:
Seems like we're doing essentially the same thing in 3 places:
* load an env and json.Unmarshal it into the passed in arg
* BindOptions on a flagset
* complete the options

So it seems like this could be abstracted to:
```
type foo interface {
  BindOptions(flags flag.FlagSet)
  Complete(unusedArgs []string)
}

func Bar(f *foo, var string) {
  json.Unmarshal(os.LookupEnv(var), foo)
  fs := flag.NewFlagSet()
  foo.BindOptions(fs)
  fs.Parse()
  foo.Complete(fs.Args())
}
```

Then `gcsUpload`, `sidecar` and `initupload` can each just do `Bar(&gcsUpload.options{}, gcsUpload.JSONConfigEnvVar)`

Maybe there are other places we use this pattern too?

Reply:
Not every binary takes arguments, so not every options structure has a `Complete()` -- how would you apply this pattern to aggregated options like in `sidecar`? Today we have:

```go
fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
gcsupload.BindOptions(options.GcsOptions, fs)
wrapper.BindOptions(options.WrapperOptions, fs)
fs.Parse(os.Args[1:])
options.GcsOptions.Complete(fs.Args())
```

The approach you outlined seems reasonable but as soon as you have multiple `foo`s that do or do not `Complete()` is it still useful?

Reply:
Seems like sidecar follows the same pattern to me:

```
type Options struct {
	GcsOptions     *gcsupload.Options `json:"-"`
	WrapperOptions *wrapper.Options   `json:"-"`
}

func (o *Options) Bind(flags) {
 o.GcsOptions.Bind(flags)
 o.Wrapper.Bind(flags)
}

func (o *Options) Complete(args) {
  o.GcsOptions.Complete(args)
}
```

Is there something about this that doesn't work?